In [68]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
from sklearn.linear_model import LinearRegression

In [69]:
# Let's import the dataset

url = 'https://raw.githubusercontent.com/Jenarth/SCS3250-Group-2-Final-Project/master/Final%20Dataset.csv'
df = pd.read_csv(url,index_col='Date', parse_dates=True)

In [70]:
# Let's convert weekdays into integers

weekday = {'Monday':1 , 'Tuesday':2 , 'Wednesday':3, 'Thursday':4, 'Friday':5, 'Saturday':6, 'Sunday':7}
df.Weekday = [weekday[item] for item in df.Weekday]
df.head()

Year  Month year_month  Day  Weekday  Holiday  Max Temp (°C)  \
Date                                                                       
2017-03-01  2017      3    2017-03    1        3    False           13.1   
2017-03-02  2017      3    2017-03    2        4    False           -1.3   
2017-03-03  2017      3    2017-03    3        5    False           -4.3   
2017-03-04  2017      3    2017-03    4        6    False           -6.5   
2017-03-05  2017      3    2017-03    5        7    False           -0.4   

            Min Temp (°C)  Mean Temp (°C)  Monthly Mean Temp (°C)  ...  \
Date                                                               ...   
2017-03-01           -1.9             5.6                0.116129  ...   
2017-03-02           -6.4            -3.9                0.116129  ...   
2017-03-03           -8.1            -6.2                0.116129  ...   
2017-03-04          -11.9            -9.2                0.116129  ...   
2017-03-05          -10.6            -5.5                0.116129  ...   

            Lakeshore West Ridership  Milton Ridership  \
Date                                                     
2017-03-01                    243965            115524   
2017-03-02                    251245            115976   
2017-03-03                    233488            104386   
2017-03-04                     87326              1518   
2017-03-05                     62331              1234   

            Richmond Hill Ridership  Stouffville Ridership  \
Date                                                         
2017-03-01                    41759                  65258   
2017-03-02                    41975                  64723   
2017-03-03                    38029                  58671   
2017-03-04                        0                   1792   
2017-03-05                        0                   1135   

            Union Station Ridership  Total GO Ridership  Outlier  \
Date                                                               
2017-03-01                   709669             1538388    False   
2017-03-02                   710233             1550282    False   
2017-03-03                   645929             1413668    False   
2017-03-04                   102982              262464    False   
2017-03-05                    62998              171582    False   

            Overall Mean     2SD_above     2SD_below  
Date                                                  
2017-03-01  1.684403e+06  1.885102e+06  1.483705e+06  
2017-03-02  1.667167e+06  1.877280e+06  1.457055e+06  
2017-03-03  1.513443e+06  1.722796e+06  1.304090e+06  
2017-03-04  3.363528e+05  5.477078e+05  1.249978e+05  
2017-03-05  2.407631e+05  3.932424e+05  8.828385e+04  

[5 rows x 28 columns]

In [72]:
# Build linear regression model using Weekday and Precipitation as predictors
# Split data into predictors X and output Y

predictors = ['Weekday','Total Precip (mm)']
X= df[predictors]
y=df['Union Station Ridership'].values.reshape(-1,1)

In [74]:
# Initialise and fit model
lm = LinearRegression()
model = lm.fit(X, y)

Instead of calulcating the values for alpha and betas, we have to call .intercept for alpha, and .coef_ for an array with our coefficients.

In [75]:
print(f'alpha = {model.intercept_}')
print(f'betas = {model.coef_}')

alpha = [966292.03323924]
betas = [[-107545.69052024   -1043.23097587]]


We can predict the values by using .predict()

In [76]:
y_pred = model.predict(X)
y_pred

array([[624459.51172257],
       [535900.6249631 ],
       [428563.58063804],
       ...,
       [532875.25513309],
       [427416.02656458],
       [319766.01294676]])

In [77]:
new_X = [[7, 0.2]]
print(model.predict(new_X))

[[213263.55340238]]


Now that we have our predicting model, let's randomly pick a day and assess how far (or close) our model is from reality.

In [79]:
df[['Weekday','Total Precip (mm)','Union Station Ridership']].tail(15)

Weekday  Total Precip (mm)  Union Station Ridership
Date                                                           
2020-02-15        6                0.4                   113314
2020-02-16        7                0.4                    90869
2020-02-17        1                1.0                    84243
2020-02-18        2                8.6                   773600
2020-02-19        3                0.0                   795204
2020-02-20        4                1.0                   800284
2020-02-21        5                0.2                   703123
2020-02-22        6                0.2                   127953
2020-02-23        7                0.8                    93485
2020-02-24        1                1.1                   795210
2020-02-25        2                0.2                   736552
2020-02-26        3               11.8                   727372
2020-02-27        4                3.1                   703693
2020-02-28        5                1.1                   704527
2020-02-29        6                1.2                    98604

In [81]:
# Let's look at the 18th of February
prediction_1 = [[2, 8.6]]
print(model.predict(prediction_1))

[[742228.8658063]]


The model predicted 742,228 riderships while the actual number was 773,600: the prediction was pretty close.

In [84]:
# Let's look at the 26th of February 2020
prediction_2 = [[3, 11.8]]
print(model.predict(prediction_2))

[[631344.83616329]]


The model predicted 631,344 riderships while the actual number was 727,372. The prediction here was not as accurate.